pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go

```diff
@@ -261,3 +323,105 @@ func (m *NoopManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]b
 func (m *NoopManager) GetTrustAnchorsBySigner(signerName string, labelSelector *metav1.LabelSelector, allowMissing bool) ([]byte, error) {
 	return nil, fmt.Errorf("ClusterTrustBundle projection is not supported in static kubelet mode")
 }
+
+// LazyInformerManager decides whether to use the noop or the actual manager on a call to
+// the manager's methods.
+// We cannot determine this upon startup because some may rely on the kubelet to be fully
+// running in order to setup their kube-apiserver.
+type LazyInformerManager struct {
+	manager           Manager
+	managerLock       sync.Mutex
+	client            clientset.Interface
+	cacheSize         int
+	contextWithLogger context.Context
+	logger            logr.Logger
+}
+
+func NewLazyInformerManager(ctx context.Context, kubeClient clientset.Interface, cacheSize int) Manager {
+	return &LazyInformerManager{
+		client:            kubeClient,
+		cacheSize:         cacheSize,
+		contextWithLogger: ctx,
+		logger:            klog.FromContext(ctx),
+		managerLock:       sync.Mutex{},
+	}
+}
+
+func (m *LazyInformerManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]byte, error) {
+	if m.manager == nil {
+		// the above check is thread-unsafe but we don't want to sync over it. It should
+		// be attempted again in a locked context
+		m.determineManager()
+	}
+	return m.manager.GetTrustAnchorsByName(name, allowMissing)
+}
+
+func (m *LazyInformerManager) GetTrustAnchorsBySigner(signerName string, labelSelector *metav1.LabelSelector, allowMissing bool) ([]byte, error) {
+	if m.manager == nil {
+		// the above check is thread-unsafe but we don't want to sync over it. It should
+		// be attempted again in a locked context
+		m.determineManager()
+	}
+	return m.manager.GetTrustAnchorsBySigner(signerName, labelSelector, allowMissing)
+}
+
+type managerConstructor func(ctx context.Context, informerFactory informers.SharedInformerFactory, cacheSize int, cacheTTL time.Duration) (Manager, error)
+
+func (m *LazyInformerManager) determineManager() {
+	m.managerLock.Lock()
+	defer m.managerLock.Unlock()
+	if m.manager != nil {
+		return
+	}
+
+	managerSchema := map[schema.GroupVersion]managerConstructor{
+		certificatesv1alpha1.SchemeGroupVersion: NewAlphaInformerManager,
+		certificatesv1beta1.SchemeGroupVersion:  NewBetaInformerManager,
+	}
+
+	kubeInformers := informers.NewSharedInformerFactoryWithOptions(m.client, 0)
+
+	var clusterTrustBundleManager Manager
+	for _, gv := range []schema.GroupVersion{certificatesv1beta1.SchemeGroupVersion, certificatesv1alpha1.SchemeGroupVersion} {
+		ctbAPIAvailable, err := clusterTrustBundlesAvailable(m.client, gv)
```

Comment:
similar comment about treating a NotFound on the group/version and a missing individual resource consistently